.github/workflows/test.yml

```
@@ -0,0 +1,33 @@
+name: Run tests
+
+on: [push, pull_request]
+
+jobs:
+  test:
+    name: Test on Python ${{ matrix.python-version }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12', '3.13']
+      fail-fast: false
+
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install docker compose
+        run: |
+          sudo apt update
+          sudo apt install docker-compose
+      - name: Install hatch
+        run: pip install hatch
+      - name: Set up hatch project environment
+        run: hatch env create
+        env:
+          HATCH_PYTHON: ${{ matrix.python-version }}
+      - name: Run tests
+        run: hatch run test
+      - name: Run pre-commit checks
+        run: hatch run fmt
```

```suggestion
      - name: Run pre-commit checks
        run: hatch run fmt
      - name: Run tests
        run: hatch run test
```

I think it would be better to fail on linter changes first to catch those early.